File: recruit.ai (4)/src/services/emailService.ts
import imap from 'imap-simple';
import { simpleParser } from 'mailparser';
import { settings } from '../core/config';
import { queryOne, runQuery } from '../db/session';
import { processResumeLogic } from './resumeProcessor';
import fs from 'fs';
import path from 'path';
import nodemailer from 'nodemailer';

export const RESUMES_RAW_DIR = path.join(process.cwd(), 'resumes_raw');
if (!fs.existsSync(RESUMES_RAW_DIR)) {
  fs.mkdirSync(RESUMES_RAW_DIR);
}

export const getHREmail = async () => {
  const row = await queryOne("SELECT value FROM settings WHERE key = 'hr_email'");
  return row?.value || settings.HR_EMAIL;
};

export const sendNotificationEmail = async (subject: string, body: string) => {
  const emailSetting = await queryOne("SELECT value FROM settings WHERE key = 'email_notifications'");
  if (emailSetting?.value === 'false') {
    console.log("[EMAIL NOTIFICATION] Email notifications are disabled in settings. Skipping.");
    return;
  }

  const hrEmail = await getHREmail();
  console.log(`[EMAIL NOTIFICATION] To: ${hrEmail}, Subject: ${subject}`);
  
  const rawUser = settings.EMAIL_USER;
  const rawPass = settings.EMAIL_PASS;

  if (!rawUser || !rawPass || rawPass === 'YOUR_GMAIL_APP_PASSWORD') {
    console.log("[EMAIL NOTIFICATION] Email credentials not configured. Skipping real email send.");
    return;
  }

  const userEmail = rawUser.trim().replace(/^["']|["']$/g, '');
  const cleanPassword = rawPass.trim().replace(/\s+/g, '').replace(/^["']|["']$/g, '');

  try {
    const transporter = nodemailer.createTransport({
      host: settings.SMTP_HOST,
      port: settings.SMTP_PORT,
      secure: settings.SMTP_PORT === 465,
      auth: {
        user: userEmail,
        pass: cleanPassword,
      },
    });

    await transporter.sendMail({
      from: `"Recruit.AI" <${userEmail}>`,
      to: hrEmail,
      subject: `[Recruit.AI] ${subject}`,
      text: body,
    });

    console.log(`[EMAIL NOTIFICATION] Successfully sent email to ${hrEmail}`);
  } catch (err) {
    console.error(`[EMAIL NOTIFICATION] Failed to send email:`, err);
  }
};

const sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

export const testEmailConnection = async () => {
  const rawUser = settings.EMAIL_USER;
  const rawPass = settings.EMAIL_PASS;

  if (!rawUser || !rawPass || rawPass === 'YOUR_GMAIL_APP_PASSWORD') {
    return { success: false, error: "Email credentials not configured or placeholder detected." };
  }

  // Clean credentials: remove whitespace and potential quotes
  const userEmail = rawUser.trim().replace(/^["']|["']$/g, '');
  const cleanPassword = rawPass.trim().replace(/\s+/g, '').replace(/^["']|["']$/g, '');

  const maskedEmail = userEmail.length > 5 
    ? `${userEmail.substring(0, 3)}...${userEmail.substring(userEmail.indexOf('@') - 2)}`
    : '***';
  const maskedPass = cleanPassword.length > 2 
    ? `${cleanPassword[0]}...${cleanPassword[cleanPassword.length-1]}` 
    : '***';
  console.log(`[INGESTION] Testing connection for user: ${maskedEmail}, host: ${settings.EMAIL_HOST}, port: ${settings.EMAIL_PORT}`);
  console.log(`[INGESTION] Diagnostic: Email length=${userEmail.length}, Password length=${cleanPassword.length}, MaskedPass=${maskedPass}`);

  if (settings.EMAIL_HOST?.toLowerCase().includes('smtp')) {
    return { success: false, error: `Invalid EMAIL_HOST: ${settings.EMAIL_HOST}. You are using an SMTP host for an IMAP connection. Please use imap.gmail.com instead.` };
  }

  if (cleanPassword.length !== 16 && settings.EMAIL_HOST?.includes('gmail')) {
    console.warn(`[INGESTION] Warning: App Password length is ${cleanPassword.length}, expected 16 for Gmail.`);
  }

  const config = {
    imap: {
      user: userEmail,
      password: cleanPassword,
      host: settings.EMAIL_HOST,
      port: settings.EMAIL_PORT,
      tls: true,
      tlsOptions: { rejectUnauthorized: false },
      authTimeout: 30000,
      connTimeout: 30000,
    }
  };

  try {
    const connection = await imap.connect(config);
    connection.end();
    return { success: true, message: "Connection successful!" };
  } catch (err: any) {
    const errMsg = err.message || String(err);
    console.error(`[INGESTION] Connection test failed:`, errMsg);
    if (err.textCode === 'AUTHENTICATIONFAILED' || errMsg.includes('Invalid credentials') || errMsg.includes('Authentication failed')) {
      return { 
        success: false, 
        error: "Authentication failed. \n" +
               "1. If using Gmail, you MUST use a 16-character 'App Password' (not your regular password).\n" +
               "2. Ensure IMAP is enabled in Gmail settings (Forwarding and POP/IMAP tab).\n" +
               "3. Check if your email address is correct."
      };
    }
    return { success: false, error: `Connection failed: ${errMsg}` };
  }
};

export const performIngestion = async (retryCount = 3) => {
  const rawUser = settings.EMAIL_USER;
  const rawPass = settings.EMAIL_PASS;

  if (!rawUser || !rawPass || rawPass === 'YOUR_GMAIL_APP_PASSWORD') {
    console.log("[INGESTION] Email credentials not configured or placeholder detected. Skipping ingestion.");
    return { success: false, message: "Email not configured" };
  }

  const hrEmail = await getHREmail();

  // Clean credentials: remove whitespace and potential quotes
  const userEmail = rawUser.trim().replace(/^["']|["']$/g, '');
  const cleanPassword = rawPass.trim().replace(/\s+/g, '').replace(/^["']|["']$/g, '');

  const maskedEmail = userEmail.length > 5 
    ? `${userEmail.substring(0, 3)}...${userEmail.substring(userEmail.indexOf('@') - 2)}`
    : '***';
  const maskedPass = cleanPassword.length > 2 
    ? `${cleanPassword[0]}...${cleanPassword[cleanPassword.length-1]}` 
    : '***';
  console.log(`[INGESTION] Initializing ingestion for user: ${maskedEmail}, host: ${settings.EMAIL_HOST}, port: ${settings.EMAIL_PORT}`);
  console.log(`[INGESTION] Diagnostic: Email length=${userEmail.length}, Password length=${cleanPassword.length}, MaskedPass=${maskedPass}`);

  if (settings.EMAIL_HOST?.toLowerCase().includes('smtp')) {
    console.error(`[INGESTION] Error: Invalid EMAIL_HOST: ${settings.EMAIL_HOST}. Using SMTP host for IMAP connection.`);
    return { success: false, message: "Invalid email host configuration" };
  }

  if (cleanPassword.length !== 16 && settings.EMAIL_HOST?.includes('gmail')) {
    console.warn(`[INGESTION] Warning: App Password length is ${cleanPassword.length}, expected 16 for Gmail.`);
  }

  const config = {
    imap: {
      user: userEmail,
      password: cleanPassword,
      host: settings.EMAIL_HOST,
      port: settings.EMAIL_PORT,
      tls: true,
      tlsOptions: { rejectUnauthorized: false },
      authTimeout: 30000,
      connTimeout: 30000,
    }
  };

  let connection: any;
  let attempt = 0;

  while (attempt < retryCount) {
    try {
      attempt++;
      console.log(`[INGESTION] Connection attempt ${attempt}/${retryCount} to ${settings.EMAIL_HOST} as ${userEmail}...`);
      connection = await imap.connect(config);
      break; // Success
    } catch (err: any) {
      const errMsg = err.message || String(err);
      console.error(`[INGESTION] Connection attempt ${attempt} failed:`, errMsg);
      
      if (err.textCode === 'AUTHENTICATIONFAILED' || errMsg.includes('Invalid credentials') || errMsg.includes('Authentication failed')) {
        return { 
          success: false, 
          error: "Authentication failed. This is usually caused by an incorrect password or missing App Password.\n" +
                 "1. Ensure you are using a 16-character 'App Password' from your Google Account settings (Security -> 2-Step Verification -> App Passwords).\n" +
                 "2. Ensure IMAP is enabled in your Gmail settings: https://mail.google.com/mail/u/0/#settings/fwdandpop\n" +
                 "3. Verify your email address is correct in the settings."
        };
      }
      
      if (attempt >= retryCount) {
        return { success: false, error: `Failed to connect after ${retryCount} attempts: ${errMsg}` };
      }
      const delay = Math.pow(2, attempt) * 1000;
      console.log(`[INGESTION] Retrying in ${delay / 1000}s...`);
      await sleep(delay);
    }
  }

  try {
    const limitRow = await queryOne("SELECT value FROM settings WHERE key = 'email_ingestion_limit'");
    const limit = parseInt(limitRow?.value || '5', 10);

    await connection.openBox('INBOX');

    const searchCriteria = ['UNSEEN', ['FROM', hrEmail]];
    const fetchOptions = { bodies: ['HEADER', 'TEXT', ''], struct: true };

    const messages = await connection.search(searchCriteria, fetchOptions);
    const messagesToProcess = messages.slice(0, limit);

    console.log(`[INGESTION] Found ${messages.length} messages, processing up to ${limit}.`);

    let processedCount = 0;
    for (const message of messagesToProcess) {
      const messageId = message.attributes.uid.toString();
      
      try {
        const processed = await queryOne("SELECT message_id FROM processed_emails WHERE message_id = ?", [messageId]);
        if (processed) {
          console.log(`[INGESTION] Message ${messageId} already processed. Skipping.`);
          continue;
        }

        const all = message.parts.find((part: any) => part.which === '');
        if (!all) {
          console.warn(`[INGESTION] Message ${messageId} has no body part. Skipping.`);
          continue;
        }
        
        const parsed = await simpleParser(all.body);
        const attachments = parsed.attachments || [];
        const processedAttachments = [];

        console.log(`[INGESTION] Processing message ${messageId} from ${parsed.from?.text}. Found ${attachments.length} attachments.`);

        for (const attachment of attachments) {
          const ext = path.extname(attachment.filename || '').toLowerCase();
          if (['.pdf', '.docx'].includes(ext)) {
            const filename = `email_${messageId}_${attachment.filename}`;
            const filepath = path.join(RESUMES_RAW_DIR, filename);
            fs.writeFileSync(filepath, attachment.content);
            
            processedAttachments.push({
              path: filepath,
              type: "resume",
              content: attachment.content,
              ext: ext
            });
          }
        }

        if (processedAttachments.length > 0) {
          const primary = processedAttachments[0];
          await processResumeLogic(primary.path, primary.content, primary.ext, processedAttachments);
          processedCount++;
        }

        await runQuery("INSERT INTO processed_emails (message_id, processed_at) VALUES (?, ?)", [messageId, new Date().toISOString()]);
      } catch (msgError) {
        console.error(`[INGESTION] Failed to process message ${messageId}:`, msgError);
      }
    }

    connection.end();
    await runQuery("UPDATE settings SET value = ? WHERE key = 'last_sync_at'", [new Date().toISOString()]);
    console.log(`[INGESTION] Finished processing. Successfully ingested ${processedCount} resumes.`);
    return { success: true, processedCount };
  } catch (error: any) {
    if (connection) connection.end();
    console.error("Email Ingestion Error during processing:", error);
    return { success: false, error: String(error) };
  }
};

File: recruit.ai (4)/app/services/email_service.py
import os
import datetime
import smtplib
import asyncio
try:
    from imap_tools import MailBox, AND
    IMAP_AVAILABLE = True
except ImportError:
    IMAP_AVAILABLE = False
    print("imap-tools package not found.")

from concurrent.futures import ThreadPoolExecutor
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from pathlib import Path
RESUMES_RAW_DIR = Path("resumes_raw")
RESUMES_RAW_DIR.mkdir(exist_ok=True)

from app.core.config import settings

HR_SENDER_EMAIL = settings.HR_EMAIL

async def send_notification_email(subject: str, body: str):
    """Sends an email notification using SMTP."""
    email_user = settings.EMAIL_USER
    email_pass = settings.EMAIL_PASS
    smtp_host = settings.SMTP_HOST
    smtp_port = settings.SMTP_PORT
    
    if not email_user or not email_pass:
        print("SMTP credentials not set, skipping notification email.")
        return

    try:
        msg = MIMEMultipart()
        msg['From'] = email_user
        msg['To'] = HR_SENDER_EMAIL
        msg['Subject'] = subject
        msg.attach(MIMEText(body, 'plain'))

        server = smtplib.SMTP(smtp_host, smtp_port)
        server.starttls()
        server.login(email_user, email_pass)
        server.send_message(msg)
        server.quit()
        print(f"Notification email sent: {subject}")
    except Exception as e:
        print(f"Failed to send notification email: {e}")

executor = ThreadPoolExecutor(max_workers=3)

import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Throttling limit
INGEST_LIMIT = 5

def _sync_ingest():
    from app.services.resume_processor import process_resume_logic_sync
    from app.db.session import get_db_conn
    if not IMAP_AVAILABLE:
        logger.error("IMAP service unavailable - imap-tools not installed")
        return {"success": False, "message": "IMAP service unavailable"}
    
    email_user = settings.EMAIL_USER
    email_pass = settings.EMAIL_PASS
    email_host = settings.EMAIL_HOST
    
    if not email_user or not email_pass:
        logger.error("Email credentials not set in environment variables")
        return {"success": False, "message": "Email credentials not set"}

    processed_count = 0
    conn = get_db_conn()
    cursor = conn.cursor()
    
    try:
        logger.info(f"Connecting to IMAP server: {email_host}")
        # Added timeout to prevent hanging indefinitely
        with MailBox(email_host, timeout=30).login(email_user, email_pass) as mailbox:
            logger.info(f"Successfully logged in as {email_user}. Fetching unread emails from {HR_SENDER_EMAIL} (Limit: {INGEST_LIMIT})")
            
            # Only fetch unread emails from the specific HR sender, with a limit for throttling
            emails = mailbox.fetch(AND(seen=False, from_=HR_SENDER_EMAIL))
            
            for i, msg in enumerate(emails):
                if i >= INGEST_LIMIT:
                    logger.info(f"Reached ingestion limit of {INGEST_LIMIT} emails. Throttling...")
                    break
                
                # Check if email already processed using message_id
                cursor.execute("SELECT 1 FROM processed_emails WHERE message_id = ?", (msg.message_id,))
                if cursor.fetchone():
                    logger.info(f"Email with Message-ID {msg.message_id} already processed. Skipping.")
                    mailbox.flag(msg.uid, "\\Seen", True)
                    continue

                logger.info(f"Processing email: {msg.subject} from {msg.from_}")
                email_processed_successfully = False
                all_attachments = []
                
                for att in msg.attachments:
                    ext = Path(att.filename).suffix.lower()
                    if ext in [".pdf", ".docx", ".txt", ".jpg", ".png"]:
                        # Include message_id hash and a unique ID for absolute uniqueness
                        import uuid
                        msg_id_short = "".join(filter(str.isalnum, msg.message_id))[:10]
                        timestamp = int(datetime.datetime.now().timestamp())
                        unique_id = str(uuid.uuid4())[:8]
                        filename = f"email_{msg_id_short}_{timestamp}_{unique_id}_{Path(att.filename).name}"
                        filepath = RESUMES_RAW_DIR / filename
                        try:
                            with open(filepath, "wb") as f:
                                f.write(att.payload)
                            
                            logger.info(f"Saved attachment: {filename}")
                            
                            # Determine file type
                            file_type = "other"
                            if "resume" in att.filename.lower() or "cv" in att.filename.lower():
                                file_type = "resume"
                            elif "cover" in att.filename.lower() or "letter" in att.filename.lower():
                                file_type = "cover_letter"
                            elif ext in [".pdf", ".docx"] and not any(a["type"] == "resume" for a in all_attachments):
                                file_type = "resume" # Default first doc to resume if not specified
                            
                            all_attachments.append({
                                "path": filepath,
                                "content": att.payload,
                                "ext": ext,
                                "type": file_type,
                                "name": att.filename
                            })
                            
                            processed_count += 1
                            email_processed_successfully = True
                        except Exception as e:
                            logger.error(f"Failed to save attachment {att.filename}: {e}")
                            
                if email_processed_successfully:
                    # Process all attachments for this email together
                    # We'll pick the one marked as 'resume' as the primary
                    primary_resume = next((a for a in all_attachments if a["type"] == "resume"), all_attachments[0])
                    process_resume_logic_sync(primary_resume["path"], primary_resume["content"], primary_resume["ext"], all_attachments)
                    
                    # Record email as processed
                    cursor.execute("INSERT INTO processed_emails (message_id, processed_at) VALUES (?, ?)", 
                                 (msg.message_id, datetime.datetime.now().isoformat()))
                    conn.commit()
                
                mailbox.flag(msg.uid, "\\Seen", True)
                
        logger.info(f"Ingestion completed. Processed {processed_count} resumes.")
        return {"success": True, "processedCount": processed_count}
    except Exception as e:
        error_msg = str(e).lower()
        logger.error(f"IMAP Ingestion Error: {str(e)}")
        
        # Check for specific IMAP error conditions
        if "authentication failed" in error_msg:
            return {"success": False, "message": "IMAP Authentication Failed. Please check your email credentials."}
        elif "timed out" in error_msg or "timeout" in error_msg:
            return {"success": False, "message": "IMAP Connection Timed Out. The server is taking too long to respond. Please try again later."}
        elif "quota exceeded" in error_msg or "mailbox full" in error_msg or "over quota" in error_msg:
            return {"success": False, "message": "IMAP Mailbox Full. Your email account has reached its storage limit. Please clear some space."}
        elif "connection" in error_msg or "network" in error_msg:
            return {"success": False, "message": "IMAP Connection Error. Could not connect to the email server. Please check your host settings and network connection."}
        
        return {"success": False, "message": f"An unexpected email ingestion error occurred: {str(e)}"}
    finally:
        conn.close()

async def perform_ingestion():
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(executor, _sync_ingest)
